Add option to disable redundant 132B in log lines

yaten2302 opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This PR adds an option to disable the resource attribute in logs.

<!-- Issue number if applicable -->
#### Link to tracking issue
Fixes #13869

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Added unit tests in - `TestCreateLogger()`.

<!--Please delete paragraphs that you did not use before submitting.-->

bogdandrutu commented on service/telemetry/internal/migration/v0.3.0.go:
```diff
@@ -162,6 +162,14 @@ type LogsConfigV030 struct {
+	// If false, resource attributes neither be added to logs exported through the LoggerProvider nor
+	// to the logs written to stdout/stderr.
+	// (default = true)
+	ResourceAsZapFields bool `mapstructure:"resource_as_zap_field"`
```
Not sure where implementation specific things should go. @open-telemetry/collector-approvers

yaten2302 commented:
@jade-guiton-dd , wanted to ask here, that currently, `TestCreateLoggerWithResource` test is failing in `logger_test.go`.
If I explicitly mention like this:
```go
{
			name: "auto-populated fields only",
			buildInfo: component.BuildInfo{
				Command: "mycommand",
				Version: "1.0.0",
			},
			resourceConfig: map[string]*string{},
			wantFields: map[string]string{
				string(semconv.ServiceNameKey):       "mycommand",
				string(semconv.ServiceVersionKey):    "1.0.0",
				string(semconv.ServiceInstanceIDKey): "",
			},
			cfg: &Config{
				Logs: LogsConfig{
					Level:               zapcore.InfoLevel,
					Encoding:            "json",
					ResourceAsZapFields: true,
				},
				Resource: map[string]*string{},
			},
		},
```
Then, it'll work fine, but, it's not working if I don't add this field - `ResourceAsZapFields`. How, can I make it's value as `true` by default?
I tried using the `defaults` lib as well - "github.com/mcuadros/go-defaults". But, still the default value isn't being set as true. I'm trying to figure this out, if you've any suggestions/hints for this, kindly share :)

jade-guiton-dd commented:
The default configuration for the telemetry factory is in `service/telemetry/otelconftelemetry/factory.go`, you'll want to modify the `createDefaultConfig` function to set the new field.

yaten2302 commented:
@jade-guiton-dd , I've added the required field in the `createDefaultConfig()` func in `factory.go`. But, if we check in the `TestCreateLoggerWithResource()` func, we're passing our own config (currently that code is commented for testing purpose). So, if we pass our own config, then the default one will be overwritten, right? Because, if I test this func by passing - `createDefaultConfig` to `createLogger` func, then the 2 tests are passing and others which are failing are because it's meant to be (because of `wantFields`).

In short, the as far as I've observed here is that, this - `if cfg.Logs.ResourceAsZapFields && res != nil && len(res.Attributes()) > 0 {` (which I added) is working fine, but, I'm in a dilemma that if we pass a custom config, then what should be the behaviour? Should it merge the default and the custom config or the default config shouldn't be taken into account (while passing the default one)? Since, the current behaviour in my latest commit is this only. The default config is getting ignored if we pass the custom one, and if we pass the default config (in which we've `ResourceAsZapFields` as true by default), it's working perfectly fine.

atoulme commented on .chloggen/main.yaml:
```diff
@@ -0,0 +1,25 @@
+component: "otelconftelemetry"
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Add support to disable resource attribute"
```
```suggestion
note: "Add support to disabling adding resource attributes as zap fields in internal logging"
```

yaten2302 commented on .chloggen/main.yaml:
```diff
@@ -0,0 +1,25 @@
+change_type: "enhancement"
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: "otelconftelemetry"
```
@jade-guiton-dd , @atoulme , could you kindly have a look at this, ig after this we'll be good to merge this PR 👀 ?

yaten2302 replied:
Ig this is fixed, could someone resolve this?

jade-guiton-dd commented on service/telemetry/otelconftelemetry/logger.go:
```diff
@@ -55,7 +55,7 @@ func createLogger(
 	// even ones exported through the core that wraps the LoggerProvider,
 	// meaning that the attributes would be exported twice.
-	if res != nil && len(res.Attributes()) > 0 {
+	if cfg.Logs.DisableResourceAttributes && res != nil && len(res.Attributes()) > 0 {
```
Since the new flag is meant to _disable_ resource attributes, there should be a `!` here

jade-guiton-dd commented:
The contrib-tests are failing for everyone, it's not related to your PR

yaten2302 commented:
Ok, got this 👍 
Should I update the changelog file and update the component name to this - `pkg/service`?
@jade-guiton-dd

jade-guiton-dd reviewed (approved):
Looks good. Suggested a few documentation changes to make it extra clear what this applies to.
